text/0000-signals-in-lwc-templates.md

```diff
@@ -0,0 +1,196 @@
+---
+title: Supporting Signals in LWC Templates
+status: DRAFT
+created_at: 2023-11-15
+updated_at: 2023-11-15
+champion: Caridy & Dale 
+pr: https://github.com/salesforce/lwc-rfcs/pull/
+---
+
+# Supporting Signals in LWC Templates
+
+## Summary
+
+This RFC proposes an enhancement to the rendering mechanism in Lightning Web Components (LWC) to support Signals-based objects in LWC Templates, this is inspired by similar mechanisms in libraries such as Preact, vue, etc. The goal is to provide a flexible mechanism to integrate stores or signals into the template in such a way that the rehydration of the UI is still possible even when the user is not relying on the observable membrane provided by LWC. The introduction of Signals in LWC opens the door for more ergonomic methods of observing state changes in the component and allows developers to bring their own data stores if they adhere to a simple protocol involving `.value`, and `.subscribe` properties. This change can also accelerate the work on performance optimizations within LWC Rendering process, especially when rehydrating the template, potentially reducing the need for full component re-rendering through a more efficient subscription mechanism.
+
+## Motivation
+
+The current reactivity model in LWC relies solely in the observable membrane, via proxies, which are tightly coupled with LWC internals. This coupling leads to certain limitations:
+
+1. __Limited External Store Interaction__: LWC components cannot directly interact with external state management systems without going through LWC-specific APIs. This restricts the framework's flexibility and developer freedom to integrate with the broader JavaScript ecosystem.
+1. __Complexity and Overhead__: The use of a membrane based on proxies introduces additional complexity and overhead in component design and state management, from read-and-write to read-only objects, it can be challenging for developers to debug and reason about data flow and updates.
+1. __Performance Constraints__: The current model relies on components to re-render fully to reflect state changes, which can be inefficient, especially for large and complex component trees.
```

Comment:
I see this as orthogonal, especially because for a v1, we will probably just re-render the whole template anyway (similar to the current system). We'd have to implement [fine-grained reactivity](https://github.com/salesforce/lwc/issues/3624) to change that.

Reply:
well, I think the hint of a `.value` in the template, it is certainly the difference. If we were to continue doing what we are doing, we could get far, but because we don't know what is reactive and what's not reactive, I suspect we might have perf challenges.

Reply:
You know what's reactive, but that doesn't mean you necessarily want fine-grained reactivity at the level of every observable property. Even Solid batches them together in some cases.

Without Signals, I think we can batch at a very coarse level, e.g. for individual items in a `for:each` iteration, and get a lot of bang for our buck.

My main point is that I wouldn't sell Signals based on perf. I'd sell it based on DX.